### ISSUE TITLE
PVC delete command hangs for a PVC belonging to readyToUse=false volumesnapshot. (PVC remains in Terminating status for hours).

### ISSUE THREAD

**Issue by kulkarnicr** [IBM/ibm-spectrum-scale-csi#254]
**Describe the bug**
While trying to delete PVC that is assigned to a snapshot in readyToUse=false state,
I observed that PVC delete command hangs.  PVC remains in Terminating status for hours (e.g. 14hrs in my case).

**To Reproduce**

1. Create PVC (don't wait for PVC to go in Bound status)
2. Create volumesnapshot before PVC is Bound
3. volumesnapshot shows readyToUse=false
4. Wait for PVC Bound status
5. Verify snapshot shows readyToUse=true. This is not happening. Snapshot remains in readyToUse=false for hours. This problem is tracked through issue #253.
6. In order to get rid of this hanging snapshot (readyToUse=false), I tried to Delete this snapshot
7. Verify volumesnapshot is deleted => it got deleted.
8. Try deleting PVC that was used for snapshot
9. Verify PVC is deleted
=> observed that PVC deletion command didn't complete (hangs). PVC was not deleted, it remained in Terminating status forever.

Steps with command execution output:

1. Create PVC (don't wait for PVC to go in Bound status)

```
[root@gn1 2020_07_07-03:04:44 test2]$ kubectl -n ibm-spectrum-scale-csi-driver create -f pvc_10mb-1001.yaml
persistentvolumeclaim/pvc-1001 created
[root@gn1 2020_07_07-03:05:09 test2]$
```

2. Create volumesnapshot before PVC is Bound

```
[root@gn1 2020_07_07-03:05:09 test2]$ kubectl -n ibm-spectrum-scale-csi-driver create -f volumesnapshot-1001.yaml
volumesnapshot.snapshot.storage.k8s.io/snap-1001 created
[root@gn1 2020_07_07-03:05:11 test2]$
[root@gn1 2020_07_07-03:05:13 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get pvc pvc-1001
NAME       STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS                       AGE
pvc-1001   Pending                                      ibm-spectrum-scale-csi-1k-inodes   8s
[root@gn1 2020_07_07-03:05:17 test2]$
```

3. volumesnapshot shows readyToUse=false

```
[root@gn1 2020_07_07-03:05:18 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get volumesnapshot snap-1001
NAME        READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS   SNAPSHOTCONTENT   CREATIONTIME   AGE
snap-1001   false        pvc-1001                                          snapclass1                                       11s
[root@gn1 2020_07_07-03:05:22 test2]$
```

4. Wait for PVC Bound status

```
[root@gn1 2020_07_07-03:05:23 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get pvc pvc-1001 -w
NAME       STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS                       AGE
pvc-1001   Pending                                      ibm-spectrum-scale-csi-1k-inodes   18s
pvc-1001   Pending   pvc-39a30862-6692-427e-8e51-64d39f1d4d19   0                         ibm-spectrum-scale-csi-1k-inodes   19s
pvc-1001   Bound     pvc-39a30862-6692-427e-8e51-64d39f1d4d19   11Mi       RWX            ibm-spectrum-scale-csi-1k-inodes   19s
^C[root@gn1 2020_07_07-03:05:31 test2]$
```

5. Verify snapshot shows readyToUse=true
This is not happening. Snapshot remains in readyToUse=false for hours. This problem is tracked through issue #253.

```
[root@gn1 2020_07_07-03:05:35 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get volumesnapshot snap-1001
NAME        READYTOUSE   SOURCEPVC   SOURCESNAPSHOTCONTENT   RESTORESIZE   SNAPSHOTCLASS   SNAPSHOTCONTENT   CREATIONTIME   AGE
snap-1001   false        pvc-1001                                          snapclass1                                       29s
[root@gn1 2020_07_07-03:05:40 test2]$
```

In order to get rid of this hanging snapshot (readyToUse=false), I tried following.
6. Delete this snapshot

```
[root@gn1 2020_07_07-03:08:15 test2]$ kubectl -n ibm-spectrum-scale-csi-driver delete volumesnapshot snap-1001
volumesnapshot.snapshot.storage.k8s.io "snap-1001" deleted
[root@gn1 2020_07_07-03:08:18 test2]$
```

7. Verify volumesnapshot is deleted => it got deleted.

```
[root@gn1 2020_07_07-03:08:19 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get volumesnapshot snap-1001
Error from server (NotFound): volumesnapshots.snapshot.storage.k8s.io "snap-1001" not found
[root@gn1 2020_07_07-03:08:21 test2]$
```

8. Try deleting PVC that was used for snapshot. If you see the below output, "kubectl ... delete pvc pvc-1001" command hangs.
It doesn't return back to the prompt.

```
[root@gn1 2020_07_07-03:08:45 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get volumesnapshot snap-1001
Error from server (NotFound): volumesnapshots.snapshot.storage.k8s.io "snap-1001" not found
[root@gn1 2020_07_07-03:08:52 test2]$
[root@gn1 2020_07_07-03:08:52 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get pvc pvc-1001
NAME       STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                       AGE
pvc-1001   Bound    pvc-39a30862-6692-427e-8e51-64d39f1d4d19   11Mi       RWX            ibm-spectrum-scale-csi-1k-inodes   3m45s
[root@gn1 2020_07_07-03:08:54 test2]$
[root@gn1 2020_07_07-03:08:55 test2]$ kubectl -n ibm-spectrum-scale-csi-driver delete pvc pvc-1001
persistentvolumeclaim "pvc-1001" deleted

```

9. Verify PVC is deleted
=> observed that PVC deletion command didn't complete (hangs). PVC was not deleted, it remained in Terminating status forever.

```
[root@gn1 2020_07_07-03:58:32 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get pvc pvc-1001
NAME       STATUS        VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                       AGE
pvc-1001   Terminating   pvc-39a30862-6692-427e-8e51-64d39f1d4d19   11Mi       RWX            ibm-spectrum-scale-csi-1k-inodes   53m
[root@gn1 2020_07_07-03:58:37 test2]$
```

10. I've observed one another similar PVC sitting in Terminating status for 14+hours.

```
[root@gn1 2020_07_07-03:58:38 test2]$ kubectl -n ibm-spectrum-scale-csi-driver get pvc scale-fset-pvc-10mb-998
NAME                      STATUS        VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                       AGE
scale-fset-pvc-10mb-998   Terminating   pvc-f47e2ee7-560f-4b1b-a34f-dd18cde5b7b8   11Mi       RWX            ibm-spectrum-scale-csi-1k-inodes   15h
[root@gn1 2020_07_07-03:58:44 test2]$
```

**Expected behavior**
PVC deletion command shouldn't hang. 
PVC shouldn't remain in Terminating status. 
PVC should be deleted.

**Environment**

# Deployment
```
[root@gn2 2020_07_06-11:26:38 ~]$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:52:00Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
The connection to the server localhost:8080 was refused - did you specify the right host or port?
[root@gn2 2020_07_06-11:26:44 ~]$
[root@gn2 2020_07_06-11:26:45 ~]$ rpm -qa | grep gpfs
gpfs.docs-5.0.5-0.noarch
gpfs.callhome-ecc-client-5.0.5-0.noarch
gpfs.gss.pmsensors-5.0.5-0.el7.x86_64
gpfs.kafka-5.0.5-0.x86_64
gpfs.base-5.0.5-0.x86_64
gpfs.msg.en_US-5.0.5-0.noarch
gpfs.gskit-8.0.55-12.x86_64
gpfs.java-5.0.5-0.x86_64
gpfs.adv-5.0.5-0.x86_64
gpfs.license.adv-5.0.5-0.x86_64
gpfs.gss.pmcollector-5.0.5-0.el7.x86_64
gpfs.librdkafka-5.0.5-0.x86_64
gpfs.gpl-5.0.5-0.noarch
gpfs.compression-5.0.5-0.x86_64
gpfs.crypto-5.0.5-0.x86_64
[root@gn2 2020_07_06-11:26:49 ~]$
```
**Screenshots**
-

**Additional context**
While above PVC deletion was hanging, I could create/delete new PVC/snapshot.
Logs are collected and stored on test s/m under /root/issues/pvc_deletion_hanging_in_Terminating_status/ .

**Comment by smitaraut** [comment]
This is related to issue #253. PVC cannot be deleted while a snapshot from it is in progress. And the snapshot is considered to be in progress until it gets into readyToUse state. In this case the problem was that the snapshot was never getting into readyToUse state (due to issue #253 ) and hence PVC could not be deleted.

This should be fixed with latest snapshotter and snapshot-controller images. Please try with canary or v2.2.0-rc1 images.

**Comment by kulkarnicr** [comment]
> PVC cannot be deleted while a snapshot from it is in progress. And the snapshot is considered to be in progress until it gets into readyToUse state. In this case the problem was that the snapshot was never getting into readyToUse state (due to issue #253 ) and hence PVC could not be deleted.

sounds logical. I'll retry this once issue #253 is fixed and verified.

**Comment by kulkarnicr** [comment]
the recreate steps for this issue were dependent on failure observed in issue #253
( issue #253 + delete snapshot + delete pvc ).

however, issue #253 is fixed with latest builds. I verified and closed it. This automatically resolves current issue #254. So, closing it.
